cvmfs/catalog_balancer_impl.h

```diff
@@ -0,0 +1,189 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_CATALOG_BALANCER_IMPL_H_
+#define CVMFS_CATALOG_BALANCER_IMPL_H_
+
+#include "catalog_balancer.h"
+
+#include <inttypes.h>
+
+#include <cassert>
+#include <string>
+#include <vector>
+
+#include "catalog_mgr.h"
+#include "directory_entry.h"
+#include "hash.h"
+#include "logging.h"
+
+
+using namespace std;  // NOLINT
+
+namespace catalog {
+
+template <class CatalogMgrT>
+DirectoryEntryBase
+CatalogBalancer<CatalogMgrT>::MakeEmptyDirectoryEntryBase(
+    string name, uid_t uid, gid_t gid) {
+  shash::Algorithms algorithm = catalog_mgr_->spooler_->GetHashAlgorithm();
+  shash::Any file_hash(algorithm);
+  shash::HashString("", &file_hash);
+
+  DirectoryEntryBase deb;
+  deb.name_ = NameString(name);
+  deb.mode_ = S_IFREG | S_IRWXU;
+  deb.checksum_ = file_hash;
+  deb.mtime_ = time(NULL);
+  deb.uid_ = uid;
+  deb.gid_ = gid;
+  return deb;
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::AddCatalogMarker(string path) {
+  XattrList xattr;
+  DirectoryEntry parent;
+  bool retval;
+  retval = catalog_mgr_->LookupPath(PathString(path), kLookupSole, &parent);
+  assert(retval);
+  DirectoryEntryBase cvmfscatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfscatalog", parent.uid(),
+                                           parent.gid());
+  DirectoryEntryBase cvmfsautocatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfsautocatalog", parent.uid(),
+                                           parent.gid());
+  string relative_path = path.substr(1);
+  catalog_mgr_->AddFile(cvmfscatalog, xattr, relative_path);
+  catalog_mgr_->AddFile(cvmfsautocatalog, xattr, relative_path);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::Balance(catalog_t *catalog) {
+  if (catalog == NULL) {
+    // obtain a copy of the catalogs
+    vector<catalog_t*> catalogs = catalog_mgr_->GetCatalogs();
+    // we need to reverse the catalog list in order to analyze the
+    // last added ones first. This is necessary in the weird case the child
+    // catalog's underflow provokes an overflow in the father
+    reverse(catalogs.begin(), catalogs.end());
+    for (unsigned i = 0; i < catalogs.size(); ++i)
+      Balance(catalogs[i]);
+    return;
+  }
+  string catalog_path = catalog->path().ToString();
+  virtual_node_t root_node(catalog_path, catalog_mgr_);
+  root_node.ExtractChildren(catalog_mgr_);
+  // we have just recursively loaded the entire virtual tree!
```

Comment:
Totally true. It was changed in the CatalogBalancer, but not in the WritableCatalogManager